test_util/test_runner.py

```diff
@@ -18,94 +17,158 @@
 logging.basicConfig(format=LOGGING_FORMAT, level=logging.DEBUG)
 log = logging.getLogger(__name__)
 
-
-@contextmanager
-def remote_port_forwarding(tunnel, host_list, remote_key_path):
-    """Forwards port 5000 from each host in in host_list to port 5000
-    on tunnnel.host. This allows docker on hosts in host_list to pull from
-    an insecure registry on tunnel host
-
-    Args:
-        tunnel: SSHTunnel instance connecting to host to be forwarded to
-        host_list: list of host names to forward port 5000 from
-        remote_key_path: arbitary path on tunnel.host for storing key
-    """
-    tunnel.write_to_remote(tunnel.ssh_key_path, remote_key_path)
-    tunnel.remote_cmd(['chmod', '600', remote_key_path])
-    cmd_list = [
-            'ssh', '-i', remote_key_path,
-            '-l', tunnel.ssh_user,
-            '-T', '-n', '-N',
-            '-oBatchMode=yes',
-            '-oGatewayPorts=yes',
-            '-oStrictHostKeyChecking=no',
-            '-oUserKnownHostsFile=/dev/null',
-            '-oExitOnForwardFailure=yes',
-            '-R', '5000:127.0.0.1:5000']
-    p_list = []
-    log.info('Starting proxy of port 5000 from {} to {}'.format(host_list, tunnel.host))
-    for host in host_list:
-        host_cmd_list = cmd_list + [host]
-        p = Process(target=tunnel.remote_cmd, args=(host_cmd_list,))
-        p.start()
-        p_list.append(p)
-    yield
-    for p in p_list:
-        p.terminate()
+REGISTRY_DOCKERFILE = """FROM registry:2
+ADD include/certs /certs
+ENV REGISTRY_HTTP_TLS_CERTIFICATE=/certs/client.cert
+ENV REGISTRY_HTTP_TLS_KEY=/certs/client.key
+"""
 
 
 def pkg_filename(relative_path):
     return pkg_resources.resource_filename(__name__, relative_path)
 
 
-def prepare_test_registry(tunnel, test_dir):
-    """Transfer resources and issues commands on host to build test app,
-    host it on a docker registry, and prepare the integration_test container
+def deploy_testing_registry(tunnel, test_dir, agent_list):
+    """Method does the following:
+    1. utilizes the localhost to generate self signed certs
+    2. creates custom registry image with certs baked in
+    3. deploys custom registry on test host
+    4. Adds CA cert to test host and all nodes in agent list
+    5. Transfers test_server resouces to test host and builds
+        test_server docker image
+    6. Pushes test_server image to custom registry on test host
 
     Args:
-        tunnel: SSHTunnel instance
-        test_dir (str): path to be used for landing docker archives
+        tunnel: instance of SSHTunnel connecting to test host
+        test_dir: directory on test host where files can land
+        agent_list: list of IPs that might need to pull from
+            this registry
     """
+    log.info('Generating Self-Signed Certificate...')
+    check_call(['openssl', 'version'])  # Check that openssl is accessible
+    temp_dir = tempfile.mkdtemp()
```

Comment:
should make sure to clean this up since it's on the TeamCity host (with + TemporaryDirectory in python preferrably)